telegram/ext/dispatcher.py

```diff
@@ -604,27 +578,33 @@ def remove_error_handler(self, callback):
         """
         self.error_handlers.pop(callback, None)
 
-    def dispatch_error(self, update, error):
+    def dispatch_error(self, update, error, promise=None):
         """Dispatches an error.
 
         Args:
             update (:obj:`str` | :class:`telegram.Update` | None): The update that caused the error
             error (:obj:`Exception`): The error that was raised.
+            promise (:class:`telegram.utils.Promise`, optional): The promise whose pooled function
+                raised the error.
 
         """
+        async_params = None if not promise else (promise.args, promise.kwargs)
+
         if self.error_handlers:
             for callback, run_async in self.error_handlers.items():
                 if run_async:
                     if self.use_context:
                         self.run_async(callback,
                                        update,
-                                       CallbackContext.from_error(update, error, self),
+                                       CallbackContext.from_error(update, error, self,
+                                                                  async_params=async_params),
                                        update=update)
                     else:
                         self.run_async(callback, self.bot, update, error, update=update)
                 else:
                     if self.use_context:
-                        callback(update, CallbackContext.from_error(update, error, self))
+                        callback(update, CallbackContext.from_error(update, error, self,
+                                                                    async_params=async_params))
                     else:
                         callback(self.bot, update, error)
```

Comment:
This look like hell. (no offence) 
And all of these indents are to make a one liner in the case of use_context.
Also, it is intended for backward compatibility for non-context-callbacks.

I suggest instead:
```python
                cb_ctx = CallbackContext.from_error(update, error, self, async_params=async_params)
                if run_async:
                    if self.use_context:
                        self.run_async(callback,update, cb_ctx, update=update)
                    else:
                        self.run_async(callback, self.bot, update, error, update=update)
                else:
                    if self.use_context:
                        callback(update, cb_ctx)
                    else:
                        callback(self.bot, update, error)
```

This "costs" time only for those who hadn't upgraded to context base callbacks yet (they will have to by next major release!) but looks so much readable and sane.
Once non-context-callbacks will be removed from the code, this will look even cleaner.

Reply:
I had to revert the order of the `run_async` and `use_context` tests as CC will raise an error when being created with a non-context dispatcher